obspy/core/utcdatetime.py

```diff
@@ -343,32 +348,49 @@ def _set(self, **kwargs):
         microsecond = kwargs.get('microsecond', self.microsecond)
         julday = kwargs.get('julday', None)
         if julday:
-            self.timestamp = UTCDateTime(year=year, julday=julday, hour=hour,
-                                         minute=minute, second=second,
-                                         microsecond=microsecond).timestamp
+            self._ns = UTCDateTime(year=year, julday=julday, hour=hour,
+                                   minute=minute, second=second,
+                                   microsecond=microsecond)._ns
         else:
-            self.timestamp = UTCDateTime(year, month, day, hour, minute,
-                                         second, microsecond).timestamp
+            self._ns = UTCDateTime(year, month, day, hour, minute,
+                                   second, microsecond)._ns
 
-    def _from_datetime(self, dt, ms=0):
+    def _get_ns(self):
+        return self.__ns
+
+    def _set_ns(self, value):
+        if not isinstance(value, int):
+            raise TypeError('nanoseconds must be set as int/long type')
+        self.__ns = value
+
+    _ns = property(_get_ns, _set_ns)
+
+    def _from_datetime(self, dt):
         """
         Use Python datetime object to set current time.
 
         :type dt: :class:`datetime.datetime`
         :param dt: Python datetime object.
-        :type ms: float
-        :param ms: extra seconds to add to current UTCDateTime object.
         """
         # see datetime.timedelta.total_seconds
         try:
             td = (dt - TIMESTAMP0)
         except TypeError:
             td = (dt.replace(tzinfo=None) - dt.utcoffset()) - TIMESTAMP0
-        self.timestamp = (td.microseconds + (td.seconds + td.days * 86400) *
-                          1000000) / 1000000.0 + ms
+        self._ns = \
+            (td.days * 60 * 60 * 24 + td.seconds) * 1000000000 + \
```

Comment:
It would be more readable, and less bug prone, to replace all these zeros with scientific notation: e.g., `1e9`.